source/Mlos.Model.Services.Client/OptimizerServiceEncoderDecoder.cs

```diff
@@ -0,0 +1,419 @@
+                {
+                    instance.GuestSubgrids.Add(EncodeSubgridJoin(subgrid));
+                }
+            }
+
+            foreach (var dimension in hypergrid.Dimensions)
+            {
+                instance.Dimensions.Add(EncodeDimension(dimension));
+            }
+
+            instance.Name = hypergrid.Name;
+
+            return instance;
+        }
+
+        public OptimizerService.OptimizationProblem EncodeOptimizationProblem(OptimizationProblem problem)
+        {
+            OptimizerService.OptimizationProblem instance = new OptimizerService.OptimizationProblem();
+            instance.ParameterSpace = EncodeHypergrid(problem.ParameterSpace);
+            instance.ObjectiveSpace = EncodeHypergrid(problem.ObjectiveSpace);
+            foreach (var objective in problem.Objectives)
+            {
+                instance.Objectives.Add(EncodeObjective(objective));
+            }
+
+            if (problem.ContextSpace == null)
+            {
+                instance.EmptyContext = new OptimizerService.Empty();
+            }
+            else
+            {
+                instance.ContextSpace = EncodeHypergrid(problem.ContextSpace);
+            }
+
+            return instance;
+        }
+
+        public OptimizerService.Objective EncodeObjective(OptimizationObjective objective)
+        {
+            return new OptimizerService.Objective
+            {
+                Minimize = objective.Minimize,
+                Name = objective.Name,
+            };
+        }
+
+        public OptimizerService.GuestSubgrid EncodeSubgridJoin(SubgridJoin subgridJoin)
+        {
+            return new OptimizerService.GuestSubgrid
+            {
+                ExternalPivotDimension = EncodeDimension(subgridJoin.OnExternalJoin),
+                Subgrid = EncodeHypergrid(subgridJoin.Subgrid),
+            };
+        }
+
+        private OptimizerService.DimensionType DimensionToGrpcType(DimensionTypeName typeName)
+        {
+            switch (typeName)
+            {
+                case DimensionTypeName.CategoricalDimension:
+                    return OptimizerService.DimensionType.Categorical;
+                case DimensionTypeName.ContinuousDimension:
+                    return OptimizerService.DimensionType.Continuous;
+                case DimensionTypeName.DiscreteDimension:
+                    return OptimizerService.DimensionType.Discrete;
+                case DimensionTypeName.OrdinalDimension:
+                    return OptimizerService.DimensionType.Ordinal;
+                default:
+                    System.Console.Error.WriteLine("Invalid type provided to OptimizerServiceEncoderDecoder.Encoder.DimensionToGrpcType(..). Type: " + typeName);
+                    return OptimizerService.DimensionType.Continuous;
+            }
+        }
+
+        public OptimizerService.Dimension EncodeDimension(IDimension dimension)
+        {
+            if (dimension is EmptyDimension dimension1)

```

Comment:
How does this syntax work? What is the reason for the consecutive suffixes?

Reply:
This is equivalent to:
if (dimension is EmptyDimension){
   var dimension1 = (EmptyDimension) dimension
   ....
}
(Not exactly equivalent as C# prevents reuse of the same name (ex: dimension1) multiple times. Otherwise I would use the same local variable name multiple times)

VisualStudio suggested this.

Reply:
Neat :) 
consider renaming dimension1 -> emptyDimension; dimension2 -> continuousDimension; etc.
Also consider adding a comment explaining the syntax a bit - I asked around and found more people unaware of this neat little feature.

Reply:
[zaporter/grpc 630599a6] Improved readability of if (var is Type varname) and removed old comment